Initial goals and requirements. (#1)
did-coop/wallet-attached-storage-spec: spec.md

### Overview

### Goals and Requirements

This storage specification is intended to support the following goals and 
requirements.

#### Local-first and Offline capable storage

Users and apps need to be able to use (provision, set up, and start reading and 
writing to) storage spaces without being connected to the internet.

#### Storage and sharing of public, permissioned, and private encrypted data

> Although the local-first offline functionality is necessary, writing data to 

Reviewer: Since this is more important, let's just lead with it instead of the less important thing?

stable internet-accessible URIs for the purposes of sharing them is one of the 
primary use cases of this specification.

* A user needs to be able to write data (that is intended to be world-readable) 
  to a cloud-accessible URI, and be able to send that URI to intended recipients
  via any out of band mechanism such as email, chat, and so on.

* User needs to be able to change or revoke permissions at any point after 

>   sharing. Note that changed permissions apply only to subsequent operations 

Author: Means it's not retroactive, we cannot reach into the storage spaces of clients and somehow remove data that they already got.

>   sharing. Note that changed permissions apply only to subsequent operations 

Author: I'm ok with removing that whole sentence, this was just trying to proactively address questions by people who are like "why doesn't your spec do the impossible and solve DRM"?

>   sharing. Note that changed permissions apply only to subsequent operations 

Reviewer: ty I understand better now.

  (this spec is not intended to solve the general problem of DRM).

* The sharing and permission system needs to be primarily based on authorization
  capabilities (zcaps), but it needs to also support storage-side access control 
  list or similar functionality (even if only as a way for an authorized client 
  to receive an appropriate zcap)

* The sharing mechanism needs to be flexible and granular. For example, a given
  data resource needs to be: world-readable, or readable by groups or categories, 
  or by only those possessing the required authorization capabilities, or by 
  noone except the author or controller, etc

* Advanced sharing conditions are also desireable (such as "this share expires
  after X amount of time" or "this is a one-time share, and will expire after
  the first successful read request")

#### Stored data is opaque to the storage provider

* Stored data needs to be opaque to the storage provider. That is, unless a user 
explicitly authorizes it (via the sharing and permission mechanisms), storage 

> providers need to be unable to read the data stored on their servers in an 

Reviewer: ```suggestion
providers should not read the data stored on their servers in an 
```

The stricter language seems not possible without really advanced crypto. Sometimes it seems we want service provider to read e.g. ACL data out of a space, and I don't think there is any meachnism we can introduce that prevents them from accessing the data without explicit authorization.

Author: This is just a requirement for client-side encryption. Which can be done by existing crypto.

Author: "need to be unable to" means "this spec must make fully client-side-encrypted spaces possible, though not required"

Author: Rephrased.
